### DISCUSSION TITLE
Custom EP3321/40 project

### DISCUSSION THREAD

**Discussion by samdurnxn** [TillFleisch/ESPHome-Philips-Smart-Coffee#71] [Support for EPxxxx]
Hello,
Thanks for your works.
I purchase an EP3321/40 and start my own project to manage the coffee machine. I use a D1 mini and a local webserver to handle the front page with websocket between.
As I'd like to improve my knowledge in coding with esp8266, so I 'm not using ESPHome.
I make a little debug webpage to see every command which is on RX/TX link, and I found almost every command but have issue for 3 things, the selection of AquaClean, Calc/NCalc and pre-ground coffee.
As mentionned in the user guide, for these functionalites, you have to do a long press (> 3s). When I done it manually, I saw many times the same command so I think it's the right one.

For the pre-ground coffee, I try differents ways to simulate a long press but it doesn't work.
I try : 
- a loop to write the command on serial -> no effect
- ticker to send the command every 50ms -> coffee machine accept command but it's like I press the button very quickly not a long press.

Can you resume the way you have handled the long press? I'm not familiar with espHome code, sorry.

Another question, can you tell me if you think this model normally have a command to start it without purge water ? In my first analyse, when I start the machine manually after a long period, and after a short one, I have the same commands. Maybe for the starting, it's now the main program which decides to purge water or not ?

So far, I found these commands : 
```
powerOn :         {0xd5, 0x55, 0x0a, 0x01, 0x05, 0x01, 0x16, 0x00, 0x00, 0x00, 0x1f, 0x3e}
powerPurgeWater : {0xd5, 0x55, 0x01, 0x01, 0x05, 0x01, 0x16, 0x00, 0x00, 0x00, 0x34, 0x0b}
powerOff :        {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x01, 0x00, 0x00, 0x0c, 0x17}
expresso :        {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x02, 0x00, 0x00, 0x18, 0x01}
coffee :          {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x20, 0x00, 0x00, 0x31, 0x3b}
expresso lungo :  {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x08, 0x00, 0x00, 0x28, 0x30}
iced coffee :     {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x04, 0x00, 0x00, 0x30, 0x2d}
hot water :       {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x10, 0x00, 0x00, 0x18, 0x0a}
steam :           {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x00, 0x01, 0x00, 0x0c, 0x16}
coffee size :     {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x00, 0x04 ,0x00, 0x31, 0x29}
bean :            {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x00, 0x02, 0x00, 0x18, 0x03}
temperature :     {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x00, 0x08, 0x00, 0x2a, 0x38}
start/stop :      {0xd5, 0x55, 0x00, 0x01, 0x05, 0x01, 0x16, 0x00, 0x00, 0x01, 0x08, 0x1e}
```

Next steps for me : 
- find the way to handle the 3 last commands
- decode mainboard status to have right "leds" on in my webInterface
- refacto to avoid all debugs section
- create webinterface with a responsive design

**Comment by TillFleisch** [comment]
> the way you have handled the long press

The differentiation between short and long presses does not happen in the display unit but rather on the main board side. The display is simply sending, which button is currently pressed. When pressing a button, that "command" (i.e. information about which button is currently pressed) is sent repeatedly until the user releases the button.
Therefore, a long press is equal to sending the button message for 3 seconds or more.
As you've already figured out, sending different messages in-between results in multiple button presses instead of a single long one.

To prevent this behaviour in this codebase, forwarding of messages from the display to the Mainboard is temporarily stopped. All user input during the injection of a virtual button press is dropped. Thus, the Mainboard only sees a constant (3 second long) stream of the desired button pressed command.

> a loop to write the command on serial -> no effect

Depending on your implementation, this may or may not work. A common issue is that such a loop may generate the desired output but, that the microcontroller doesn't actually write anything to the bus because the send buffer is not flushed.


> Another question, can you tell me if you think this model normally have a command to start it without purge water ?

I don't know if the EP3321 supports power on without a cleaning cycle. If I were to take a guess, I would say yes since the machines are pretty similar. Some machines can perform an internal cleaning cycle (AFAIK), where they clean during start-up but don't dispense water via the coffee outlet.
I retrieved the different commands for power up with/without cleaning on my machine after removing power form the machine and starting it manually. Once the machine has already pre-heated, turning it off and on again (via the panel) results in a power on sequence without a cleaning cycle. My machine always performs a cleaning cycle if it has been disconnected from the grid, regardless if it's still pre-heated or not (I would guess removing power form the display results in a similar situation).

**Comment by samdurnxn** [comment]
Thanks for your answer, I retry with a simple loop but this time I add the flush function and it works !!
So, for my machine, I send the command 1 time for a short press and around 2700x for a long press.

Now, commands list is complete =)

### Commands
| Button     | Message                               |
| ---------- | ------------------------------------- |
| Power on   | `D5 55 0A 01 05 01 16 00 00 00 1F 3E` |
| PowerPurgeWater        | `D5 55 01 01 05 01 16 00 00 00 34 0B` |
| Power Off  | `D5 55 00 01 05 01 16 01 00 00 0C 17` |
| Espresso  | `D5 55 00 01 05 01 16 02 00 00 18 01` |
| Coffee   | `D5 55 00 01 05 01 16 20 00 00 31 3B` |
| Expresso lungo  | `D5 55 00 01 05 01 16 08 00 00 28 30` |
| Iced coffee  | `D5 55 00 01 05 01 16 04 00 00 30 2D` |
| Hot Water  | `D5 55 00 01 05 01 16 10 00 00 18 0A` |
| Steam   | `D5 55 00 01 05 01 16 00 01 00 0C 16` |
| Size  | `D5 55 00 01 05 01 16 00 04 00 31 29` |
| Bean   | `D5 55 00 01 05 01 16 00 02 00 18 03` |
| Temperature   | `D5 55 00 01 05 01 16 00 08 00 2A 38` |
| Play/Stop  | `D5 55 00 01 05 01 16 00 00 01 08 1E` |
| Aqua clean   | `D5 55 00 01 05 01 16 00 10 00 1C 1A` |
| Calc clean   | `D5 55 00 01 05 01 16 00 20 00 39 1B` |

### Message from mainboard
| Byte | Purpose                    | Detail                                                            |
| ---- | -------------------------- | ----------------------------------------------------------------- |
| 0    | START                      |                                                                   |
| 1    | START                      |                                                                   |
| 2    | INSTRUCTION                |                                                                   |
| 3    | Espresso-LED               | `03`/`07` - half/full brightness ; `38` - 2x espresso             |
| 4    | Iced coffee-LED             | `03`/`07` - half/full brightness  ; `38` - 2x iced coffee             |
| 5    | Espresso lungo                 | `03`/`07` - half/full brightness; `38` - 2x espresso lungo                |
| 6    | coffee / Hot water        | `03`/`07` (half/full) - Hot water selected; `38` - Coffee; `3F` - hot water and coffee led when mode select drink is active |
| 7    | Steam/ 2x Coffee    | `07` - 2x Coffee; `38` - Steam selected                    |
| 8    | Bean-LED                   | `00` - 1 LED; `38` - 2 LEDs; `3F` - 3 LEDs                        |
| 9    | Bean-LED                   | `07` - show led group; `38` - powder selected                     |
| 10   | Size-LED                   | `00` - 1 LED; `38` - 2 LEDs; `3F` - 3 LEDs; `07` -  Top LED       |
| 11   | Size-LED / Aqua Clean      | `07` - Show led group; `38` - Aqua Clean Orange   |
| 12   | AquaCLean/CalcClean         | `00` - both off; `07` blue aqua; `38` - Calc orange   |
| 13   | Temperature                 | `00` - 1 LED; `38` - 2 LEDs; `3F` - 3 LEDs                        |
| 14   | Temperature /  Water         | `07` - show led group; `38` - water empty                                                
| 15   | Waste&Warning-LED          | `07` - waste; `38` - warning sign                                 |
| 16   | Play/Pause-LED             | `07` - on                                                         |
| 17   | checksum                   |                                                                   |
| 18   | checksum                   |                                                                   |

About the starting without purge water, I will try to found the right command later. 

You can add this infos in your project, it will enrich your code to be compatible with more machines ;)
